File: src/main/java/com/prototype/architecture/layer_04_infrastructure/userrole/DbUserRoleRepository.java
package com.prototype.architecture.layer_04_infrastructure.userrole;

import com.prototype.architecture.layer_03_domain.user.UserEntity;
import com.prototype.architecture.layer_03_domain.user.UserId;
import com.prototype.architecture.layer_03_domain.userrole.UserRoleEntity;
import com.prototype.architecture.layer_03_domain.userrole.UserRolePk;
import com.prototype.architecture.layer_03_domain.userrole.UserRoleType;
import com.prototype.architecture.layer_03_domain.userrole.UserRoleRepository;
import lombok.AllArgsConstructor;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.Map;

@AllArgsConstructor
@Repository
public class DbUserRoleRepository implements UserRoleRepository<UserRoleEntity, UserRolePk> {

    private final NamedParameterJdbcTemplate jdbcTemplate;

    @Override
    public UserRoleEntity save(UserRoleEntity entity) {
        final String sql = "INSERT INTO user_role_t VALUES (:userId, :roleType, :registerTime, :updateTime)";
        final Map<String, Object> paramMap = Map.of(
                "userId", entity.getUserId().getValue(),
                "roleType", entity.getRoleType().name(),
                "registerTime", LocalDateTime.now(),
                "updateTime", LocalDateTime.now()
        );

        jdbcTemplate.update(sql, new MapSqlParameterSource(paramMap));
        return entity;
    }

    @Override
    public Iterable<UserRoleEntity> findAllByUserId(UserId userId) {
        final String sql = "SELECT * FROM user_role_t WHERE user_id = :userId";
        final Map<String, Object> paramMap = Map.of(
                "userId", userId.getValue()
        );
        final RowMapper<UserRoleEntity> rowMapper = (rs, rowNum) -> new UserRoleEntity(
                new UserId(rs.getInt(UserEntity.COLUMN_NAME_USER_ID)),
                UserRoleType.valueOf(rs.getString(UserRoleEntity.COLUMN_NAME_USER_ROLE_TYPE))
        );

        return jdbcTemplate.query(sql, paramMap, rowMapper);
    }

    @Override
    public void delete(UserRolePk pk) {
        final String sql = "DELETE FROM user_role_t WHERE user_id = :userId AND role_type = :roleType";
        final Map<String, Object> paramMap = Map.of(
                "userId", pk.getUserId().getValue(),
                "roleType", pk.getRoleType().name()
        );

        jdbcTemplate.update(sql, paramMap);
    }

}

File: src/main/kotlin/com/prototype/architecture/layer_04_infrastructure/DbUserRoleRepository.kt
package com.prototype.architecture.layer_04_infrastructure

import com.prototype.architecture.layer_03_domain.user.UserEntity
import com.prototype.architecture.layer_03_domain.user.UserId
import com.prototype.architecture.layer_03_domain.userrole.UserRoleEntity
import com.prototype.architecture.layer_03_domain.userrole.UserRolePk
import com.prototype.architecture.layer_03_domain.userrole.UserRoleRepository
import com.prototype.architecture.layer_03_domain.userrole.UserRoleType
import com.prototype.framework.configuration.jdbc.parameter.JdbcBeanPropertySqlParameterSource
import lombok.AllArgsConstructor
import org.apache.catalina.User
import org.springframework.jdbc.core.DataClassRowMapper
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.BeanPropertySqlParameterSource
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.core.simple.SimpleJdbcInsert
import org.springframework.stereotype.Repository
import java.sql.ResultSet
import java.time.LocalDateTime
import javax.sql.DataSource

@Repository
class DbUserRoleRepository(
        private val dataSource: DataSource,
        private val jdbcTemplate: NamedParameterJdbcTemplate
) : UserRoleRepository<UserRoleEntity, UserRolePk> {

    override fun save(entity: UserRoleEntity): UserRoleEntity {
        val params = JdbcBeanPropertySqlParameterSource(entity.applyForCreate())
        SimpleJdbcInsert(dataSource).withTableName(UserRoleEntity.TABLE_NAME).execute(params)
        return entity
    }

    override fun findAllByUserId(userId: UserId): List<UserRoleEntity> {
        val sql = """
            SELECT *
            FROM ${UserRoleEntity.TABLE_NAME}
            WHERE ${UserEntity.USER_ID} = :${UserEntity.USER_ID}
        """.trimIndent()
        val params = MapSqlParameterSource().apply {
            addValue(UserEntity.USER_ID, userId.value)
        }
        val rowMapper = DataClassRowMapper(UserRoleEntity::class.java)
        return jdbcTemplate.query(sql, params, rowMapper)
    }

    override fun delete(pk: UserRolePk) {
        val sql = """
            DELETE FROM ${UserRoleEntity.TABLE_NAME}
            WHERE ${UserEntity.USER_ID} = :${UserEntity.USER_ID}
            AND ${UserRoleEntity.USER_ROLE_TYPE} = :${UserRoleEntity.USER_ROLE_TYPE}
        """.trimIndent()
        val params = MapSqlParameterSource().apply {
            addValue(UserEntity.USER_ID, pk.userId.value)
            addValue(UserRoleEntity.USER_ROLE_TYPE, pk.roleType.name)
        }
        jdbcTemplate.update(sql, params)
    }
}